internals/search.py

```diff
@@ -279,3 +287,34 @@ def process_query(
 
   logging.info('features_on_page is %r', features_on_page)
   return features_on_page, total_count
+
+
+def process_negation_operations(feature_id_future_ops):
+  """ Turn all negation operations into one AND operation. E.g.
+  A - B - C -> A AND (ALL - B) AND (ALL - C) -> A AND (ALL - B - C).
+  """
+  new_future_ops = []
+  negated_id_set = None
+  for logical_op, future in feature_id_future_ops:
+    if logical_op != '-':
+      # Skip all non-negation operations.
+      new_future_ops.append((logical_op, future))
+      continue
+
+    if negated_id_set is None:
+        negated_id_set = fetch_all_feature_ids_set()
+
+    feature_ids = _resolve_promise_to_id_list(future)
+    negated_id_set.difference_update(feature_ids)
+
+  # Append the result of negation operations to the end.
+  if negated_id_set is not None:
+    new_future_ops.append(('', list(negated_id_set)))
```

Comment:
Could we append `('', negated_id_set)` without converting to a list?

Reply:
Yup we could!

Reply:
I got too many complaints from mypy. I will change it in the next PR, along with the OR logic